docs/Blog-Post.md

```diff
@@ -0,0 +1,276 @@
+
+However, Jax is a low-level framework, and we found that, by itself, it did not provide the legibility that we wanted.
+
+For us, legibility is a top concern. We therefore created two new libraries: **Haliax** and **Levanter**. Haliax is a
+named tensor library that focuses on improving the legibility and compositionality of deep learning code while still
+maintaining efficiency and scalability. Levanter is a library for training foundation models built on top of Haliax. In
+addition to the goals of legibility, efficiency, and scalability, Levanter further strives for bitwise reproducibility,
+meaning that the same code with the same data will produce the exact same result, even in the presence of preemption and
+restarts.
+
+## Haliax: Legibility via Named Tensors
+
+Haliax is a Jax library for building neural networks with named tensors, built on Jax and [Equinox](https://github.com/patrick-kidger/equinox),
+which is a neural network library built on Jax that provides a familiar, PyTorch-like module structure.
+
+Named Tensors are a powerful abstraction that allow you to give names to the axes of your tensors. These names help
+make your code more legible, more composable, and avoid bugs. In Haliax, they also form the basis of how we handle
+scale with [Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/). (See below for our tutorial)
+
+Haliax is modeled on Alexander Rush's [Tensor Considered Harmful](https://nlp.seas.harvard.edu/NamedTensor). In particular, he argues that:
+
+* Named axes are more semantically meaningful and rely less on bitrot-prone comments.
+* Broadcasting leads to unreadable strings of `view`s and `squeeze`s. (To this, I'd add implicit broadcasting is a source of bugs.)
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible for things like multi-headed attention or flexible attention masking.
+
+Let's take a look at a practical example of how Haliax can be used.
+Here's a minimal attention module implementation in Haliax. For a more detailed introduction,
+please see the [Haliax tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC).
+
+```python
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+from jax.random import PRNGKey
+import haliax as hax
+import haliax.nn as hnn
+
+# Named Axes for Tensor Dimensions
+Pos = hax.Axis("position", 1024)  # sequence
+KPos = Pos.alias("key_position")  # key sequence for attention
+Head = hax.Axis("head", 8)  # number of attention heads
+Key = hax.Axis("key", 64)  # key size
+Embed = hax.Axis("embed", 512)  # embedding size
+
+# Despite making no reference to batching or heads, this same implementation is also batch-capable and multi-headed
+# (or multi-query) and even supports attending to or from non-sequential keys (e.g. image patches)
+def attention(Key, KPos, query, key, value, mask):
+  # how similar is each query to each key
```

Comment:
Four space indent?

Reply:
i use four in general, but colab defaults to two and this is copy paste from colab.